lib-api/src/main/java/org/ergoplatform/appkit/ErgoProverBuilder.java

```diff
@@ -31,28 +33,22 @@
     ErgoProverBuilder withSecretStorage(SecretStorage storage);
 
     /**
-     * Add DHT prover input using this prover's secret as belonging to the first party (i.e. Alice).
+     * ProveDHTuple is of the form (g, h, u, v) with secret x (and unknown y), where:
+     *   h = g^y
+     *   u = g^x
+     *   v = g^xy
      *
-     * @param otherParty       address of the other party whose public key is to be included in
-     *                         Diffie-Hellman Tuple. If this prover corresponds to Alice, then `otherParty`
-     *                         is Bob.
-     * @param additionalSecret an additionalSecret necessary to construct DHT when this prover corresponds to
-     *                         a first party (e.g. Alice)
-     * @see
-     * <a href="https://github.com/ScorexFoundation/sigmastate-interpreter/blob/b3695bdb785c9b3a94545ffea506358ee3f8ed3d/sigmastate/src/test/scala/sigmastate/utxo/examples/DHTupleExampleSpecification.scala#L28">example</a>
-     */
-    ErgoProverBuilder withFirstDHTSecret(Address otherParty, GroupElement additionalSecret);
-
-    /**
-     * Add DHT prover input using this prover's secret as belonging to second party (i.e. Bob).
+     *   NOTE: We can swap x, y and obtain another tuple (g, u, h, v) with secret y (and unknown x).
+     *
+     * @param g {@Link GroupElement} instance defining g
+     * @param h {@Link GroupElement} instance defining h
+     * @param u {@Link GroupElement} instance defining u
+     * @param v {@Link GroupElement} instance defining v
+     * @param x {@Link BigInteger} instance defining x
+     * @return
      *
-     * @param otherParty address of the other party whose public key is to be included in
-     *                   Diffie-Hellman Tuple. This prover is assumed to belong to Bob, then `otherParty`
-     *                   is Alice.
-     * @see
-     * <a href="https://github.com/ScorexFoundation/sigmastate-interpreter/blob/b3695bdb785c9b3a94545ffea506358ee3f8ed3d/sigmastate/src/test/scala/sigmastate/utxo/examples/DHTupleExampleSpecification.scala#L28">example</a>
      */
-    ErgoProverBuilder withSecondDHTSecret(Address otherParty);
+    ErgoProverBuilder withDHTData(GroupElement g, GroupElement h, GroupElement u, GroupElement v, BigInteger x);
```

Comment:
Isn't it that the generator is always the same?

Reply:
In feel it should not be the same generator all the time. One variant of ErgoMix uses proveDHTuple(h, h, v, v), where h=g^y and v = g^(xy). Basically this allows us do to proveDlog using a generator different from g. That is: proveDlog_h(v) = proveDHTuple(h, h, v, v).

Since the protocol allows changing the generator, l think we should also allow it.